Add one list account manager filter to company search.

elcct opened the pull request:
### Description of change

`POST /v3/search/company/`: new filter `one_list_group_global_account_manager` was added.

This enables filtering by computed field `one_list_group_global_account_manager` that contains an id of one list account manager. If company is a subsidiary, this value will be inherited from the parent company.

### Checklist

* [ ] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-leeloo/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

reupen commented on datahub/search/dict_utils.py:
```diff
@@ -138,11 +138,30 @@ def get_dict(obj):
     return _computed_nested_dict(nested_field, id_name_dict)
 
 
+def computed_nested_contact_or_adviser_dict(nested_field):
```
Is this used anywhere?

elcct replied:
Oops, I forgot to remove. Thank you!

reupen commented on datahub/search/company/test/test_elasticsearch.py:
```diff
@@ -218,6 +218,35 @@ def test_mapping(es):
                         },
                     },
                 },
+                'one_list_group_global_account_manager': {
```
We're indexing quite a lot here, when it seems like we only need `id` to be indexed. Would be good to only index the field being filtered by, and set `index=False` on anything else.

elcct replied:
That's a very good point. I replaced this field with one that only indexes the id.

elcct commented:
The `one_list_account_owner` was already in the select-related fields list, but `global_headquarters__one_list_account_owner` wasn't, so I added it. Thank you!

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
+    )
+    def test_one_list_account_manager_with_global_headquarters_filter(
+        self,
+        selected_account_manager,
```
Doesn't seem like `selected_account_manager` is used here.

elcct replied:
Oh. Yes, I should have removed that. Thank you!

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
+            str(global_headquarters.id),
+            *{str(target_company.id) for target_company in target_companies},
+        }
+        assert search_results == expected_results
```
Would be good to add another assertion for the number of results as well (to make sure there are no duplicates).

elcct replied:
Good point! I added additional assertions now. Thank you!

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
+            (0, 1),
+            (0, 2),
+            (2, ),
+            (1, 0),
```
I'm not sure on the practical difference between `(0, 1)`, `(0, 2)`, `(1, 0)`. If we want to test with 1, 2 and 3 advisers maybe we could use `random.sample(account_managers, n)` and parametrise `n`?

elcct replied:
I changed this test to follow your suggestion. I think it is better now because it also checks the contents of `one_list_group_global_account_manager` field. Thank you!

reupen commented on datahub/search/company/test/test_elasticsearch.py:
```diff
@@ -18,7 +18,6 @@ def test_mapping(es):
         CompanySearchApp.es_model.get_write_index(),
         CompanySearchApp.name,
     )
-
```
Is this change intended?

elcct replied:
It is not. Fixed now. Thank you.

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +368,97 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
+            (1, 0),
+        ),
+    )
+    def test_one_list_account_manager_filter(self, selected_account_manager, es_with_signals):
```
Should update these to `es_with_collector` (instead of `es_with_signal`) when we rebase.

elcct replied:
I'll keep in mind!

elcct commented:
This is a good point, I missed that! I have changed the signal to also sync subsidiaries if a company is a headquarter and added test to check that `one_list_group_global_account_manager` is populated when  `one_list_account_owner` of global hq is changed.

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +369,132 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
+        assert len(response.data['results']) == 3
+        assert search_results == expected_results
+
+    def test_subsidiary_is_synced_to_es(self, es_with_signals):
```
I think we should add a specific test for the signal receiver for this (in `test_signals`) instead of testing this via the view.

elcct replied:
This is a good idea. I removed this test and created a new one in `test_signals`. Thanks!

reupen commented on datahub/search/company/signals.py:
```diff
@@ -8,10 +8,15 @@
+        sync_object_async(CompanySearchApp, instance.pk)
 
+        if instance.headquarter_type_id:
+            for company in instance.subsidiaries.all():
```
It would be good to use `sync_related_objects_async` for this (similar to other search apps):

https://github.com/uktrade/data-hub-leeloo/blob/470e8ad003057b3af655fea7ea4b7d23e8d1d6d8/datahub/search/contact/signals.py#L17-L21

elcct replied:
I did it this way first but it didn't seem to be working. Maybe I got something else wrong, because I redo it with `sync_related_objects_async` and it now works. Thank you!

reupen commented on datahub/search/company/test/test_entity_search_views.py:
```diff
@@ -361,6 +369,132 @@ def test_headquarter_type_filter(self, setup_headquarters_data, query, results):
         assert search_results == results
 
+    @pytest.mark.parametrize(
+        'n',
```
Sorry, I just meant `n` as example – would be good to use something more descriptive e.g. `num_account_managers`.

elcct replied:
Haha yes I was thinking the use of `n` was odd, but perhaps didn't have enough coffee to transform it into something like `num_account_managers`. It is fixed now. Thank you!

currycoder commented on changelog/company/add-one-list-group-account-manager-filter.api.md:
```diff
@@ -0,0 +1 @@
+`POST /v3/search/company/`: new filter `one_list_group_global_account_manager` was added.
```
Should this be `POST /v4/search/company`?

elcct replied:
Oh good catch! Updated now. Thank you.

reupen commented on datahub/search/company/test/test_signals.py:
```diff
@@ -35,3 +35,48 @@ def test_company_auto_updates_to_es(es_with_signals):
+    )
+    es_with_signals.indices.refresh()
+
+    result = get_basic_search_query(Company, same_name).execute()
```
I know the other tests in this module work this way, but rather than doing a search (which seems irrelevant for testing if the signal receiver works), I think it would be better if we fetched the document directly using something like:

```python
es_with_signals.get(
    index=CompanySearchApp.es_model.get_read_alias(),
    doc_type=CompanySearchApp.name,
    id=company.pk,
)
```

(Could add a helper function to `datahub.search.test.utils` for this as well.)

elcct replied:
I see. I would think that `get_basic_search_query` plays a role of such helper, but I can imagine how this is not the best way to do it. I'll try to modify the test though. Thank you!

reupen commented on datahub/search/test/utils.py:
```diff
@@ -62,3 +62,14 @@ def db_objects_to_es_documents(db_objects, index=None):
     )
+
+
+def get_documents_by_id(es_with_signals, app, ids):
```
`get_documents_by_ids` might be clearer (since it takes multiple IDs).

elcct commented:
I'll fix the typos whenI rebase.
